In [1]:
import pandas as pd
import matplotlib.pyplot as plt
pd.options.display.float_format = '{:.2f}'.format

filename = 'mutations.csv'
data = pd.read_csv(filename, index_col=0)
samples = data.shape[0]
data.head()

ANKRD26_GRCh37_10:27322259-27322259_Frame-Shift-Del_DEL_T-T--  \
C0                                                   0               
C1                                                   0               
NC0                                                  0               
C2                                                   0               
NC1                                                  0               

     ARID5B_GRCh37_10:63850705-63850705_Frame-Shift-Del_DEL_A-A--  \
C0                                                   0              
C1                                                   0              
NC0                                                  0              
C2                                                   0              
NC1                                                  0              

     PTEN_GRCh37_10:89717770-89717770_Frame-Shift-Del_DEL_A-A--  \
C0                                                   0            
C1                                                   0            
NC0                                                  0            
C2                                                   0            
NC1                                                  0            

     C11orf70_GRCh37_11:101937275-101937275_Frame-Shift-Del_DEL_T-T--  \
C0                                                   0                  
C1                                                   0                  
NC0                                                  0                  
C2                                                   0                  
NC1                                                  0                  

     LRRC43_GRCh37_12:122685346-122685346_Frame-Shift-Del_DEL_C-C--  \
C0                                                   0                
C1                                                   0                
NC0                                                  0                
C2                                                   0                
NC1                                                  0                

     FARP1_GRCh37_13:99092237-99092237_Frame-Shift-Del_DEL_G-G--  \
C0                                                   0             
C1                                                   1             
NC0                                                  0             
C2                                                   0             
NC1                                                  0             

     SNAPC1_GRCh37_14:62242911-62242911_Frame-Shift-Del_DEL_T-T--  \
C0                                                   0              
C1                                                   0              
NC0                                                  0              
C2                                                   0              
NC1                                                  0              

     ZC3H18_GRCh37_16:88691141-88691141_Frame-Shift-Del_DEL_C-C--  \
C0                                                   0              
C1                                                   0              
NC0                                                  0              
C2                                                   0              
NC1                                                  0              

     KIF2B_GRCh37_17:51901904-51901904_Missense-Mutation_SNP_C-C-T  \
C0                                                   0               
C1                                                   0               
NC0                                                  0               
C2                                                   0               
NC1                                                  0               

     KIF2B_GRCh37_17:51902014-51902014_Frame-Shift-Del_DEL_A-A--  ...  \
C0                                                   0            ...   
C1                                                   0            ...   
NC0                  

# Calc TP, FP, TN, FN for ALL Mutations

In [2]:
mutation_data = []

for mutation in data:
    curr_tp, curr_fp, curr_tn, curr_fn = 0, 0, 0, 0
    for index in data.index:
        if data[mutation][index] == 1:
            if index.startswith('C'):
                curr_tp += 1
            else:
                curr_fp += 1
        else: 
            if index.startswith('C'):
                curr_fn += 1
            else:
                curr_tn += 1
    mutation_data.append([mutation, curr_tp, curr_fp, curr_tn, curr_fn])

mutation_data 

tp_fp_tn_fn_chart = pd.DataFrame(columns=['Mutation', 'TP', 'FP', 'TN', 'FN'])
tp_fp_tn_fn_chart['Mutation']= [x[0] for x in mutation_data]
tp_fp_tn_fn_chart['TP']= [x[1] for x in mutation_data]
tp_fp_tn_fn_chart['FP']= [x[2] for x in mutation_data]
tp_fp_tn_fn_chart['TN']= [x[3] for x in mutation_data]
tp_fp_tn_fn_chart['FN']= [x[4] for x in mutation_data]

tp_fp_tn_fn_chart.head(10)

Mutation  TP  FP  TN  FN
0  ANKRD26_GRCh37_10:27322259-27322259_Frame-Shif...   0   1  60  49
1  ARID5B_GRCh37_10:63850705-63850705_Frame-Shift...   1   0  61  48
2  PTEN_GRCh37_10:89717770-89717770_Frame-Shift-D...   2   0  61  47
3  C11orf70_GRCh37_11:101937275-101937275_Frame-S...   2   0  61  47
4  LRRC43_GRCh37_12:122685346-122685346_Frame-Shi...   2   0  61  47
5  FARP1_GRCh37_13:99092237-99092237_Frame-Shift-...   2   0  61  47
6  SNAPC1_GRCh37_14:62242911-62242911_Frame-Shift...   3   0  61  46
7  ZC3H18_GRCh37_16:88691141-88691141_Frame-Shift...   5   0  61  44
8  KIF2B_GRCh37_17:51901904-51901904_Missense-Mut...   1   0  61  48
9  KIF2B_GRCh37_17:51902014-51902014_Frame-Shift-...   1   0  61  48

# Table of the top 10 features ranked by accuracy (with corresponding accuracy values)


In [3]:
# Accuracy = (TP + TN) / (TP + FP + TN + FN)

tp_fp_tn_fn_chart['Accuracy (%)'] = ((tp_fp_tn_fn_chart['TP'] + tp_fp_tn_fn_chart['TN']) / (tp_fp_tn_fn_chart['TP'] + tp_fp_tn_fn_chart['FP'] + tp_fp_tn_fn_chart['TN'] + tp_fp_tn_fn_chart['FN'])) *100
tp_fp_tn_fn_chart.sort_values(by='Accuracy (%)', ascending=False).head(10)


Mutation  TP  FP  TN  FN  \
169   DOCK3_GRCh37_3:51417604-51417604_Frame-Shift-D...  12   0  61  37   
478   PGM5_GRCh37_9:70993145-70993145_Missense-Mutat...  10   0  61  39   
160   ZBTB20_GRCh37_3:114058003-114058003_Frame-Shif...  11   1  60  38   
285   ACVR2A_GRCh37_2:148683686-148683686_Frame-Shif...  11   1  60  38   
448   RNF43_GRCh37_17:56435161-56435161_Frame-Shift-...  11   1  60  38   
204   UBR5_GRCh37_8:103289349-103289349_Frame-Shift-...  10   1  60  39   
259   LARP4B_GRCh37_10:890939-890939_Frame-Shift-Del...   8   0  61  41   
1081   KDM1B_GRCh37_6:18222307-18222307_3'UTR_DEL_A-A--   7   0  61  42   
585   UPF3A_GRCh37_13:115057211-115057211_Frame-Shif...   7   0  61  42   
76    TVP23C_GRCh37_17:15441469-15441469_Intron_SNP_...   7   0  61  42   

      Accuracy (%)  
169          66.36  
478          64.55  
160          64.55  
285          64.55  
448          64.55  
204          63.64  
259          62.73  
1081         61.82  
585          61.82  
76           61.82

# Find Most Useful Feature by ranking TP-FP

In [4]:
mutation_data = []

for mutation in data:
    curr_tp, curr_fp = 0, 0
    for index in data.index:
        if data[mutation][index] == 1:
            if index.startswith('C'):
                curr_tp += 1
            else:
                curr_fp += 1
    mutation_data.append([mutation, curr_tp, curr_fp])

mutation_data.sort(key=lambda x: x[1] - x[2], reverse=True)
most_useful_mutation = mutation_data[0][0]
print(f"The most useful mutation is: {most_useful_mutation}")

The most useful mutation is: DOCK3_GRCh37_3:51417604-51417604_Frame-Shift-Del_DEL_C-C--


# Find Most Useful Feature in Group A and B by ranking TP-FP

In [5]:
# Get Group A and Group B patients
group_a_patients = []
group_b_patients = []

for sample in data.index:
    if data[most_useful_mutation][sample] == 1:
        group_a_patients.append(sample)
    else:
        group_b_patients.append(sample)

max_len = max(len(group_a_patients), len(group_b_patients))
for i in range(max_len - len(group_a_patients)):
    group_a_patients.append('')
for i in range(max_len - len(group_b_patients)):
    group_b_patients.append('')

In [6]:
# Get Group A
group_a_mutation_data = []

# drop '' from group_a_patients
group_a_patients = [x for x in group_a_patients if x != '']
group_a_data = data.loc[group_a_patients]
for mutation in group_a_data:
    curr_tp, curr_fp = 0, 0
    for index in group_a_data.index:
        if group_a_data[mutation][index] == 1:
            if index.startswith('C'):
                curr_tp += 1
            else:
                curr_fp += 1
    group_a_mutation_data.append([mutation, curr_tp, curr_fp])

group_a_mutation_data.sort(key=lambda x: x[1] - x[2], reverse=True)

# drop most_useful_mutation from group_a_mutation_data to avoid duplicates
group_a_mutation_data = [x for x in group_a_mutation_data if x[0] != most_useful_mutation]

# Find the most useful mutation in Group A
group_a_most_useful_mutation = group_a_mutation_data[0][0]
print(f"The most useful mutation for group A is: {group_a_most_useful_mutation}")

The most useful mutation for group A is: UBR5_GRCh37_8:103289349-103289349_Frame-Shift-Del_DEL_T-T--


In [7]:
# Get Group B
group_b_mutation_data = []
# find the true positive and false positive for group A

#go into data and pull data for group A patients
# drop '' from group_a_patients
group_b_patients = [x for x in group_b_patients if x != '']

group_b_data = data.loc[group_b_patients]

for mutation in group_b_data:
    curr_tp, curr_fp = 0, 0
    for index in group_b_data.index:
        if group_b_data[mutation][index] == 1:
            if index.startswith('C'):
                curr_tp += 1
            else:
                curr_fp += 1
    group_b_mutation_data.append([mutation, curr_tp, curr_fp])


group_b_mutation_data.sort(key=lambda x: x[1] - x[2], reverse=True)

#drop the most useful mutation from the list
group_b_mutation_data = [x for x in group_b_mutation_data if x[0] != most_useful_mutation]


# Find the most useful mutation in Group B
group_b_most_useful_mutation = group_b_mutation_data[0][0]
print(f"The most useful mutation for group B is: {group_b_most_useful_mutation}")

The most useful mutation for group B is: ACVR2A_GRCh37_2:148683686-148683686_Frame-Shift-Del_DEL_A-A--


# Result of classifying all samples

In [8]:
# Decision Tree Classifier

# keep track of how many we get correct
total_tp, total_fp, total_tn, total_fn = 0, 0, 0, 0

for sample in data.index:
    if data.loc[sample, most_useful_mutation] == 1:
        if data.loc[sample, group_a_most_useful_mutation] == 1:
            print(f"Our tree says {sample} is classified as C")
            if sample.startswith('C'):
                total_tp += 1
            else :
                total_fp += 1
        else:
            print(f"Our tree says {sample} is classified as NC")
            if sample.startswith('NC'):
                total_tn += 1   
            else:
                total_fn += 1  
    else:
        if data.loc[sample, group_b_most_useful_mutation] == 1:
            print(f"Our tree says {sample} is classified as C")
            if sample.startswith('C'):
                total_tp += 1
            else: 
                total_fp += 1
        else:
            print(f"Our tree says {sample} is classified as NC")
            if sample.startswith('NC'):
                total_tn += 1
            else:
                total_fn += 1



Our tree says C0 is classified as NC
Our tree says C1 is classified as C
Our tree says NC0 is classified as NC
Our tree says C2 is classified as NC
Our tree says NC1 is classified as NC
Our tree says NC2 is classified as NC
Our tree says NC3 is classified as NC
Our tree says NC4 is classified as NC
Our tree says C3 is classified as NC
Our tree says NC5 is classified as NC
Our tree says NC6 is classified as NC
Our tree says NC7 is classified as NC
Our tree says C4 is classified as C
Our tree says NC8 is classified as NC
Our tree says NC9 is classified as NC
Our tree says NC10 is classified as NC
Our tree says NC11 is classified as NC
Our tree says NC12 is classified as NC
Our tree says C5 is classified as NC
Our tree says NC13 is classified as NC
Our tree says NC14 is classified as NC
Our tree says C6 is classified as NC
Our tree says C7 is classified as NC
Our tree says C8 is classified as NC
Our tree says C9 is classified as NC
Our tree says C10 is classified as NC
Our tree says NC15 

# (including a confusion matrix)

In [9]:
# make a confusion matrix with the results

print(f"Total TP: {total_tp}")
print(f"Total FP: {total_fp}")
print(f"Total TN: {total_tn}")
print(f"Total FN: {total_fn}")

confusion_matrix = pd.DataFrame({'Predicted Healthy': [total_tp, total_fp], ' Predicted Cancer': [total_fn, total_tn]}, index=['Actual Healthy', 'Actual Cancer'])
confusion_matrix



Total TP: 13
Total FP: 1
Total TN: 60
Total FN: 36


Predicted Healthy   Predicted Cancer
Actual Healthy                 13                 36
Actual Cancer                   1                 60

# Values of the metrics listed above 

In [10]:
# accuracy is how close a given set of measurements (observations or readings) are to their true value
accuracy = (total_tp + total_tn) / (total_tp + total_fp + total_tn + total_fn) * 100

# sensitivity measures how well a test can identify true positives
sensitivity = total_tp / (total_tp + total_fn) * 100

# specificity measures how well a test can identify true negatives
specificity = total_tn / (total_fp + total_tn) * 100

# percision is how close the measurements are to each other
percision = total_tp / (total_tp + total_fp) * 100

# miss rate = rate at which a system or model makes false negatives
miss_rate = total_fn / (total_fn + total_tp) * 100

# false discovery rate = percentage of times a test incorrectly identifies a "discovery" or significant result.
fdr = total_fp / (total_fp + total_tp) * 100

# false omission rate tells you how often the model mistakenly classified something as negative when it should have been positive
forr = total_fn / (total_fn + total_tn) * 100

final_table = pd.DataFrame(columns=['Metric', 'Value'])

final_table['Metric'] = ['Accuracy (%)', 'Sensitivity (%)', 'Specificity (%)', 'Percision (%)', 'Miss Rate (%)', 'FDR (%)', 'FOR (%)']

final_table['Value'] = [accuracy, sensitivity, specificity, percision, miss_rate, fdr, forr]

final_table

Metric  Value
0     Accuracy (%)  66.36
1  Sensitivity (%)  26.53
2  Specificity (%)  98.36
3    Percision (%)  92.86
4    Miss Rate (%)  73.47
5          FDR (%)   7.14
6          FOR (%)  37.50

- With a 66% accuracy rate, I personally wouldnt trust this test alone currently
- the miss rate also scares me 
    - i would rather get a false positive than a false negative